### DISCUSSION TITLE
Improved API for custom arrowheads / markers

### DISCUSSION THREAD

**Discussion by chrtze** [xyflow/xyflow#1658] [New Features]
In the next release we want to roll out a new API for customizing the markerStart and markerEnd features of the edge paths. Ideally it should be easy to use existing shapes (Arrow, Diamond, Rectangle, ...) as well as defining custom ones. The main problem is that we need to define the elements as svg markers and create unique ids for them. The same configuration object should only create one marker and the same id for all edges.

Do you have any ideas for how this api should look like? My first approach would be something like this:

```
export interface EdgeMarker {
  type: ArrowHeadType;
  color?: string;
  width?: number;
  height?: number;
  units?: string;
  orient?: string;
  strokeWidth?: number;
}
```

for the edge element the `arrowHeadType` property would be replaced by `markerStart: EdgeMarker` and `markerEnd: EdgeMarker`

**Comment by vaibhav-systango** [comment]
@chrtze I'm trying to make an ERD view and I need the custom marker edges in the following format as shown below, but I'm unable to achieve it with custom EdgeMarker config object. Can you please share any example here ?
@moklick 
![erd](https://user-images.githubusercontent.com/25978575/167409326-3e768c36-e819-4e32-85d8-d2b1ea72ac3e.png)

**Reply by vaibhav-systango** [reply]
const markerEdge = {
     source: '1',
     target: '2',
     markerStart: 'myCustomSvgMarker',
     markerEnd: { type: 'arrow', color: '#f00' }
}

how do we use this custom `myCustomSvgMarker`

**Comment by vaibhav-systango** [comment]
Here's a sample I tried doesn't works on react flow but works in a regular svg the marker start & end attributes:
<img width="936" alt="Screen Shot 2022-05-09 at 6 16 32 PM" src="https://user-images.githubusercontent.com/25978575/167412943-efadf3a4-b45d-47be-b6de-89148700ccad.png">

here's the same implemented in react flow:

<img width="849" alt="Screen Shot 2022-05-09 at 6 17 02 PM" src="https://user-images.githubusercontent.com/25978575/167412961-f56dd16e-d0fb-4b5e-b53d-b35d3d40f6b0.png">

**Reply by vaibhav-systango** [reply]
@justdvl you have any examples for this ?

**Reply by vaibhav-systango** [reply]
@chrtze 
@moklick
Any updates ? Or shall we move this to issues ?

**Reply by chrtze** [reply]
Can you please share a code example with your implementation? You would need to define the marker with the id `triangle` yourself in some svg in you app. Then you can use the id for the markerStart / markerEnd on your edge.

**Reply by vaibhav-systango** [reply]
Here's the codesandbox for same, can you help
https://codesandbox.io/s/nameless-haze-j0lx6d

check this code in ERD.tsx file
` {
    id: 'e5-6',
    source: '5',
    target: '6',
    type: 'smoothstep',
    label: 'one-to-one',
    labelBgPadding: [8, 4],
    labelBgBorderRadius: 4,
    labelBgStyle: { fill: '#FFCC00', color: '#fff', fillOpacity: 0.7 },
     markerStart: 'triangle', 
     markerEnd: 'triangle', 
  }`
  
  @chrtze @moklick

**Reply by vaibhav-systango** [reply]
![Screenshot from 2022-05-11 15-43-27](https://user-images.githubusercontent.com/25978575/167833727-cd6ad1cb-e81f-4eaa-9ed6-1fe536e50229.png)

**Reply by chrtze** [reply]
I think in your example implementation you are missing the definition of the `triangle` marker. When you provide a `markerEnd` id to the edge, it will be added to the `path` element of the edge as `marker-end` attribute. This is intended to be used as you would use it in SVG (like in the sample that you provided). When you are passing `markerEnd: 'triangle'` it will end up as `marker-end="url(#triangle)"` in the DOM, so the path will look for a marker definition (https://developer.mozilla.org/en-US/docs/Web/SVG/Element/marker) with the id `triangle`. To make it work, you would need to define the marker in a SVG outside of your React Flow application. For reference, you can check how it is implemented for the markers that are built into React Flow: https://github.com/wbkd/react-flow/blob/main/src/container/EdgeRenderer/MarkerDefinitions.tsx

Have noticed already, that this part of the API can be quite confusing, we will add a guide and improve the documentation for adding custom markers, so that it is easier to understand.

**Reply by vaibhav-systango** [reply]
can you help me with a sample example of a custom marker definition ? maybe a quick code pen link or sample code ? @chrtze @moklick
